src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
-  } else {
-    OM_PERFDATA_OP(MonExtant, inc(Atomic::load(&target->om_in_use_count)));
-  }
-
-  do {
-    int local_deflated_count;
-    if (is_global) {
-      local_deflated_count =
-          deflate_monitor_list(self, &om_list_globals._in_use_list,
-                               &om_list_globals._in_use_count,
-                               &free_head_p, &free_tail_p,
-                               &saved_mid_in_use_p);
-    } else {
-      local_deflated_count =
-          deflate_monitor_list(self, &target->om_in_use_list,
-                               &target->om_in_use_count, &free_head_p,
-                               &free_tail_p, &saved_mid_in_use_p);
+      if (ls != NULL) {
+        ls->print_cr("after handshaking: in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+        timer.start();
+      }
     }
-    deflated_count += local_deflated_count;
-
-    if (free_head_p != NULL) {
-      // Move the deflated ObjectMonitors to the global free list.
-      guarantee(free_tail_p != NULL && local_deflated_count > 0, "free_tail_p=" INTPTR_FORMAT ", local_deflated_count=%d", p2i(free_tail_p), local_deflated_count);
-      // Note: The target thread can be doing an om_alloc() that
-      // is trying to prepend an ObjectMonitor on its in-use list
-      // at the same time that we have deflated the current in-use
-      // list head and put it on the local free list. prepend_to_common()
-      // will detect the race and retry which avoids list corruption,
-      // but the next field in free_tail_p can flicker to marked
-      // and then unmarked while prepend_to_common() is sorting it
-      // all out.
-#ifdef ASSERT
-      ObjectMonitor* l_next_om = unmarked_next(free_tail_p);
-      assert(l_next_om == NULL, "must be NULL: _next_om=" INTPTR_FORMAT, p2i(l_next_om));
-#endif
 
-      prepend_list_to_global_wait_list(free_head_p, free_tail_p, local_deflated_count);
+    // After the handshake, safely free the ObjectMonitors that were
+    // deflated in this cycle.
+    size_t deleted_count = 0;
+    for (ObjectMonitor* monitor: delete_list) {
+      delete monitor;
+      deleted_count++;
+      if (self->is_Java_thread() &&
+          SafepointMechanism::should_process(self->as_Java_thread())) {
+        // A safepoint/handshake has started.
+        if (ls != NULL) {
+          timer.stop();
+          ls->print_cr("pausing deletion: deleted_count=" SIZE_FORMAT ", in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                       deleted_count, in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+        }
 
-      OM_PERFDATA_OP(Deflations, inc(local_deflated_count));
-    }
+        // Honor block request.
+        ThreadBlockInVM tbivm(self->as_Java_thread());
```

Comment:
Ditto previous comments on use of TBIVM.

Reply:
This kind of use of ThreadBlockInVM predates this changeset so while
the location is new, then code style is old, very old... I'll hold off changing
this for now.